### ISSUE TITLE
Discriminator tensor shape

### ISSUE THREAD

**Issue by edchengg** [leimao/Voice_Converter_CycleGAN#2]
Hi,

I was trying to go through the discriminator:
the input is a tensor with shape(1, 24, 128):
I am just curious about the output 'o1' shape, the output is a 6*8*1 tensor(I know this is a normal some setting for cyclegan(called patchGAN). but the Cyclegan_vc paper states FC layer at the end). Do we need to make sure the output is 1 single unit?

h1_glu [Dimension(1), Dimension(24), Dimension(64), Dimension(128)]
d1 [Dimension(1), Dimension(12), Dimension(32), Dimension(256)]
d2 [Dimension(1), Dimension(6), Dimension(16), Dimension(512)]
d3 [Dimension(1), Dimension(6), Dimension(8), Dimension(1024)]
o1 [Dimension(1), Dimension(6), Dimension(8), Dimension(1)]

btw, if the setting is LSGAN, why there is still a sigmoid activation at the end of the Discriminator?

**Comment by leimao** [comment]
Hi Yang,
Good spot.
In the CycleGAN for image conversions, the authors used the "patchGAN" you mentioned in their paper, because they were doing 2D-convolutions. I intended to do the same "patchGAN" for CycleGAN-VC but probably did not use tf.layer.dense() correctly so that a scalar was output. (For my output, I think the dimensions are o1 [Dimension(minibatch_size), Dimension(1)]). I might change it to the "patchGAN" some day if I got a chance to see if it makes any difference (You may also do the experiment and tell me the result). But using just a scalar as output seems to be working fine. (I don't have access to a linux machine right now, maybe what I said is the other way around. I will confirm this with you later.)
The sigmoid function makes the output a probability which is between 0 and 1. We are minimizing the squared error between the output probability and it supposed label (0 or 1). By convention, FNN does not contain activation function.

**Comment by edchengg** [comment]
Thank you for your reply!

I have checked your discriminator output shape it is: [batch_size, H=6, W=8, C=1]. So the output is not a single scalar.

When I use sigmoid at the end of the discriminator(pytorch implementation), the gradient vanishes very quickly.(loss of generator --> 1 and loss of discriminator --> 0). People usually remove sigmoid for LSGAN setting.

Did you add bias for the GLU layer? I see no bias in your code and according to the paper, bias are learned params.

Anyway, your code works.... I might need to further check other details with my code..

**Comment by leimao** [comment]
Thanks Yang for helping me check it out. The "patchGAN" is what I wanted to achieve.

Although I did not read through the LSGAN paper, the author only mentioned that sigmoid activation function is bad in the context of cross-entropy loss function. They did not say sigmoid activation is bad in the context of least square loss function. In their least-squared loss function, I believe they still use sigmoid activation function for the discriminator. If you check the CycleGAN for image translation paper (and their code), and the CycleGAN for voice conversion paper, there are evidences that sigmoid activation functions were used (in the PyTorch CycleGAN implementation). It does not make sense to me not to use it by the way. If you have specific examples or papers claiming that sigmoid activatoin function is also bad in the context of least square loss function, please point me to it.

GLU is just simply a Hadamard product. The biases are in the previous layers but not GLU layer.

**Comment by edchengg** [comment]
CycleGAN author: 
https://github.com/junyanz/CycleGAN/issues/70

**Comment by leimao** [comment]
Thanks. You are right. I checked the TensorFlow and PyTorch implementations, CycleGAN for image translation did not use any activation functions for least square loss in discriminators.

Did your model get improved if you remove sigmoid activation function? Was the improvement significant?

**Comment by leimao** [comment]
Hi Yang,

I checked my GLU implementation. Some of them are actually not implemented exactly the way that I thought. I might need to revise them to the actual "GLU" if I got time.

For example, the following code 

```
def residual1d_block(
    inputs, 
    filters = 1024, 
    kernel_size = 3, 
    strides = 1,
    name_prefix = 'residule_block_'):

    h1 = conv1d_layer(inputs = inputs, filters = filters, kernel_size = kernel_size, strides = strides, activation = None, name = name_prefix + 'h1_conv')
    h1_norm = instance_norm_layer(inputs = h1, activation_fn = None, name = name_prefix + 'h1_norm')
    h1_norm_gates = instance_norm_layer(inputs = h1, activation_fn = None, name = name_prefix + 'h1_norm_gates')
    h1_glu = gated_linear_layer(inputs = h1_norm, gates = h1_norm_gates, name = name_prefix + 'h1_glu')
    h2 = conv1d_layer(inputs = h1_glu, filters = filters // 2, kernel_size = kernel_size, strides = strides, activation = None, name = name_prefix + 'h2_conv')
    h2_norm = instance_norm_layer(inputs = h2, activation_fn = None, name = name_prefix + 'h2_norm')
    
    h3 = inputs + h2_norm

    return h3
```

should be revised to 

```
def residual1d_block(
    inputs, 
    filters = 1024, 
    kernel_size = 3, 
    strides = 1,
    name_prefix = 'residule_block_'):

    h1 = conv1d_layer(inputs = inputs, filters = filters, kernel_size = kernel_size, strides = strides, activation = None, name = name_prefix + 'h1_conv')
    h1_norm = instance_norm_layer(inputs = h1, activation_fn = None, name = name_prefix + 'h1_norm')

    # correction starts
    h1_gates = conv1d_layer(inputs = inputs, filters = filters, kernel_size = kernel_size, strides = strides, activation = None, name = name_prefix + 'h1_conv')
    h1_norm_gates = instance_norm_layer(inputs = h1_gates, activation_fn = None, name = name_prefix + 'h1_norm_gates')
    # correction ends

    h1_glu = gated_linear_layer(inputs = h1_norm, gates = h1_norm_gates, name = name_prefix + 'h1_glu')
    h2 = conv1d_layer(inputs = h1_glu, filters = filters // 2, kernel_size = kernel_size, strides = strides, activation = None, name = name_prefix + 'h2_conv')
    h2_norm = instance_norm_layer(inputs = h2, activation_fn = None, name = name_prefix + 'h2_norm')
    
    h3 = inputs + h2_norm

    return h3
```

Thanks again for staring at my code. Neural network is alchemy because sometimes you made mistake it still works, especially in generative models which are hard to be evaluated quantitatively.

**Comment by edchengg** [comment]
Yep!
My implementation does not work right now so I cant say anything. Thats why I am checking your code line by line to see if they are the same.

1.  I removed sigmoid function. Otherwise, my loss go to 1 for generators and 0 for discriminator immediately which indicates training fails. 

2. For GLU, some people do this for the bias.
h = tf.nn.bias_add(conv, b)*tf.sigmoid(tf.nn.bias_add(conv_gated, b_gated))
（https://github.com/dodocho/GatedCNN-based-summarizer/blob/master/encoder.py）

3. For GLU two layers: yes, there should be two conv layers. But your implementation still works... I tried both neither of them works in my case. 

Really appreciated your response and code. I learned a lot of differences between tensorflow and pytorch. For example, the instance normalization in pytorch was set to not trainable by default but in tensorflow the default is trainable.

**Comment by leimao** [comment]
Hi Yang,
I thought tf.nn.conv2d() automatically contains the bias term but it seems that it does not... I also learned a lot from you. Thank you.

**Comment by leimao** [comment]
Hi Yang,
Just let you know that [tf.layers.conv2d()](https://www.tensorflow.org/api_docs/python/tf/layers/conv2d) automatically contains bias terms. So I don't need to add bias terms to my code.

**Comment by leimao** [comment]
The GLU implementation "bug" has now been fixed. The training now of course becomes slower due to that more parameters were introduced. But it seems that the conversion becomes obvious earlier in the training compared to the previous implementation. The quality of conversion was not compromised.

**Comment by edchengg** [comment]
That’s great to hear!

**Comment by edchengg** [comment]
Could you update your loss plot when you get a chance? Thanks!

**Comment by leimao** [comment]
Yang, the training takes 400K iterations. The training is still going on at my side. I finished 100K iterations in around 15 hours, and the conversion was already great. I will update the demo and plot later.
